projects/saturn/src/main/java/io/fairspace/saturn/services/views/ViewService.java

```diff
@@ -81,55 +83,76 @@ public class ViewService {
                FILTER NOT EXISTS { ?subject fs:dateDeleted ?anyDateDeleted }
             }
             """, FS.NS));
+    public static final String USER_DOES_NOT_HAVE_PERMISSIONS_TO_READ_FACETS = "User does not have permissions to read facets";
 
     private final Config.Search searchConfig;
+    private final Config.Caches cachesConfig;
     private final ViewsConfig viewsConfig;
     private final Dataset ds;
     private final ViewStoreClientFactory viewStoreClientFactory;
     private final MetadataPermissions metadataPermissions;
-    private final LoadingCache<Boolean, List<FacetDTO>> facetsCache;
-    private final LoadingCache<Boolean, List<ViewDTO>> viewsCache;
+    private LoadingCache<Boolean, List<FacetDTO>> facetsCache;
+    private LoadingCache<Boolean, List<ViewDTO>> viewsCache;
 
-    public ViewService(Config.Search searchConfig,
+    public ViewService(Config config,
                        ViewsConfig viewsConfig,
                        Dataset ds,
                        ViewStoreClientFactory viewStoreClientFactory,
                        MetadataPermissions metadataPermissions) {
-        this.searchConfig = searchConfig;
+        this.searchConfig = config.search;
+        this.cachesConfig = config.caches;
         this.viewsConfig = viewsConfig;
         this.ds = ds;
         this.viewStoreClientFactory = viewStoreClientFactory;
         this.metadataPermissions = metadataPermissions;
-        this.facetsCache = buildFacetCache();
-        this.viewsCache = buildViewsCache();
+        if (cachesConfig.facets.enabled) {
+            this.facetsCache = buildCache(this::fetchFacets, cachesConfig.facets);
+        }
+        if (cachesConfig.views.enabled) {
+            this.viewsCache = buildCache(this::fetchViews, cachesConfig.views);
+        }
         refreshCaches();
     }
 
     public void refreshCaches() {
         log.info("Caches refreshing/warming up has been triggered");
         FilteredDatasetGraph.disableQuadPermissionCheck();
-        facetsCache.refresh(Boolean.TRUE);
-        viewsCache.refresh(Boolean.TRUE);
+        if (cachesConfig.facets.enabled) {
+            facetsCache.refresh(Boolean.TRUE);
+        }
+        if (cachesConfig.views.enabled) {
+            viewsCache.refresh(Boolean.TRUE);
+        }
         FilteredDatasetGraph.enableQuadPermissionCheck();
         log.info("Caches refreshing/warming up successfully finished");
     }
 
-    public List<FacetDTO> getCachedFacets() {
-        try {
+    public List<FacetDTO> getFacets() {
+        if (cachesConfig.facets.enabled) {
```

Comment:
what happens if cache is disable, no permission check is performed?

Reply:
not really, permission check is performed in FilteredDatasetGraph - for each particular quad retrieved from Jena

Reply:
I don't see it, not in 'fetchFacets' or in the 'calculateRead', where is it supposed to be checked?

Reply:
check `FilteredDatasetGraph` class, it is a filter on any read from Jena